crates/net/eth-wire/src/types/message.rs

```diff
@@ -214,6 +213,78 @@ impl EthMessage {
+///
+/// Implementors should ensure that encoding respects the specified size limits, either by
+/// returning an error when the limit is exceeded (`encode_max`) or by producing a truncated
+/// encoding that fits within the limit (`encode_truncate`).
+pub trait EncodableExt {
+    /// Encodes the data, ensuring that the encoded message does not exceed the specified size
+    /// limit. If the encoded size exceeds the limit, an error is returned, indicating the failure
+    /// to comply with the size constraint.
+    ///
+    /// This method is suitable for scenarios where it's critical to avoid exceeding a size limit,
+    /// such as in network communications where payload sizes are restricted.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The maximum allowed size of the data.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(Vec<u8>)` - A vector containing the encoded data if it is within the size limit.
+    /// * `Err(alloy_rlp::Error)` - An error if the encoded data exceeds the size limit.
+    fn encode_max(&self, size: usize) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error>;
+
+    /// Encodes the data, truncating the encoded message if it exceeds the specified size limit.
+    /// The resulting data will fit within the limit but may be incomplete if truncation occurs.
+    ///
+    /// When using this method with collections or "vec-like" structures, implementors should take
+    /// care to ensure the truncated encoding remains valid and interpretable, albeit potentially
+    /// incomplete. For example, truncating a list of items might result in the list containing
+    /// fewer items than intended, but each item in the list should remain individually intact
+    /// and decodable.
+    ///
+    /// # Arguments
+    ///
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A vector containing the encoded data, truncated to fit within the size limit if necessary.
+    /// Users of this method should be aware that the truncated data may not represent the complete
+    /// original dataset.
+    fn encode_truncate(&self, limit: usize) -> Vec<u8>;
+}
+
+impl<T: Encodable> EncodableExt for Vec<T> {
+    fn encode_max(&self, limit: usize) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error> {
+        let mut buffer = Vec::new();
+        for item in self {
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                return Err(alloy_rlp::Error::Custom("Size limit exceeded"));
+            }
+        }
+        Ok(buffer)
+    }
+
+    fn encode_truncate(&self, limit: usize) -> Vec<u8> {
+        let mut buffer = Vec::new();
+        for item in self {
+            let current_len = buffer.len();
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                buffer.truncate(current_len);
+                break;
+            }
+        }
+        buffer
+    }
+}
```

Comment:
this isn't quite correct because this lacks the list header. see encodeable impl for Vec

Reply:
this is less expensive if you allocate some capacity to the vector first, and then call `shrink_to_fit` on it. rust collections grow in powers of 2, so the first elements will be costly to reallocate. let's use the length approximation to call `Vec::with_capacity`. here it makes sense to use `smallvec` though I think, will go much faster if it the buffer stays on the stack.